### PR TITLE
Add a message when a variable name collides with a function name

### LINKED ISSUE #43166: Improve error message on namespace collision
```rust
struct Model;

fn model(model: Model) -> Model {
    model
}

fn main() {
    let mut model = Model;
    model = model(model);
}
```

The error message is:

```
error: expected function, found `Model`
 --> <anon>:9:13
  |
9 |     model = model(model);
  |             ^^^^^^^^^^^^
  |
note: defined here
 --> <anon>:8:9
  |
8 |     let mut model = Model;
  |         ^^^^^^^^^

error: aborting due to previous error
```

I guess, the compiler should suggest you to avoid using the same names for functions and variables.

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Fixes #43166.

### CODE REVIEW

**Comment by petrochenkov** [comment]
#43166 is usual shadowing, which happens pretty often, I don't think this is something that needs to be "fixed", especially with adding more noise into error messages.

**Comment by arielb1** [comment]
Sure, if we implement this, we want to at least check that we do indeed shadow a valid function.

File: src/librustc_typeck/check/callee.rs [hunk 1]
```diff
@@ -231,6 +231,10 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
                             err.span_note(span, "defined here");
                         }
                     }
+                    if let Def::Local(_) = def {
+                        err.help("it's better to pick different names between variables and \
+                                  functions");
```
Comment (nikomatsakis) [inline src/librustc_typeck/check/callee.rs]: I have some concerns about this message:

1. It's kind of judgemental. I personally don't consider it bad style to have a local variable with the same name as a function or method -- in fact, I consider it **good** style and I do it pretty regularly. For example, I might do `let labels = labels(...);` or `let labels = self.labels();`
2. The wording strongly suggests that there *is* a function with the same name as the local variable, but we don't check that this is the case. 
3. If we *did* do that check, then I would prefer to simplify highlight the local variable and the function and say something like "this local variable shadows the function, consider giving one of them a different name" -- i.e., just try to explain what's going wrong in an impartial way.

**Comment by nikomatsakis** [comment]
So @GuillaumeGomez I think all the comments had the same general theme -- that if you want to do this lint, it's going to need to be better integrated into name resolution so that we can give more precise advice. This will be a bit tricky, because the name resolution information we need is in resolve, but we don't uncover this error until typeck. To make it worse, by the time we see the error, we'll have forgotten what kind of expression produced the value that's being called, and just have its type (but we can re-inspect). It seems like to do it, we'd probably want to keep some information about shadowing for each path, and then after we get the error, we can inspect the expression being called, see that it's a path, and consult this side-table. That's the rough idea, anyway.

**Comment by GuillaumeGomez** [comment]
So is it worth doing it or not? I didn't really see any issue with the current error and don't think this'll be really useful.

**Comment by nikomatsakis** [comment]
This can be confusing, no doubt, but I agree it doesn't strike me as *highest* priority.

**Comment by GuillaumeGomez** [comment]
Then let's close it! :)

**Pull request closed without merging**